### ISSUE TITLE
Similarity metrics

### ISSUE THREAD

**Issue by tiagodavi** [elixir-nx/scholar#19]
I am thinking to add `Scholar.Metrics.Similarity` since there's `Scholar.Metrics.Distance` to calculate distances. 

How can I access Normal Elixir APis from defn? For example I'd like to create a MapSet:

```
@spec jaccard(Nx.t(), Nx.t()) :: Nx.t()
defn jaccard(x, y) do
    assert_same_shape!(x, y)

    MapSet.new(x)
 end
```

```
** (RuntimeError) cannot invoke MapSet.new/1 inside defn because it was not defined with defn
    (nx 0.1.0) lib/nx/defn/compiler.ex:101: Nx.Defn.Compiler.runtime_fun/4
    (nx 0.1.0) lib/nx/defn/evaluator.ex:29: Nx.Defn.Evaluator.__jit__/4
```

Do I need to use the normal def for this case?

**Comment by seanmor5** [comment]
You do not need a MapSet:

```elixir
defn jaccard(x1, x2) do
  # Convert to boolean
  x1 = Nx.not_equal(x1, 0)
  x2 = Nx.not_equal(x2, 0)

  intersection = Nx.logical_and(x1, x2)
  union = Nx.logical_or(x1, x2)
  Nx.sum(intersection) / Nx.sum(union)
end
```

**Comment by tiagodavi** [comment]
No.. it does not give the expected result.. 

```elixir 
iex(24)> x
#Nx.Tensor<
  f32[4]
  [1.0, 5.0, 3.0, 6.7]
>

iex(25)> y
#Nx.Tensor<
  f32[4]
  [5.0, 2.5, 3.1, 9.0]
>

iex(35)> Scholar.Metrics.Similarity.jaccard x, y
#Nx.Tensor<
  f32
  1.0
>
```

I was expecting something close to: 0.142857142857143

**Comment by tiagodavi** [comment]
```python
x = set([1.0, 5.0, 3.0, 6.7])
y = set([5.0, 2.5, 3.1, 9.0])

len(x.intersection(y)) / len(x.union(y))
```

**Comment by tiagodavi** [comment]
![Screen Shot 2022-08-12 at 19 42 03](https://user-images.githubusercontent.com/1058350/184453602-c341cc57-60ec-4bd1-8d4d-1499ffc96256.png)

**Comment by polvalente** [comment]
I think something along these lines is the way:

```elixir
iex(2)> x = Nx.tensor([1, 5, 3, 6.7])
#Nx.Tensor<
  f32[4]
  [1.0, 5.0, 3.0, 6.699999809265137]
>
iex(3)> y = Nx.tensor([5, 2.5, 3.1, 9])
#Nx.Tensor<
  f32[4]
  [5.0, 2.5, 3.0999999046325684, 9.0]
>
iex(4)> Nx.new_axis(y, 1)
#Nx.Tensor<
  f32[4][1]
  [
    [5.0],
    [2.5],
    [3.0999999046325684],
    [9.0]
  ]
>
iex(5)> Nx.new_axis(y, 1) |> Nx.equal(x)
#Nx.Tensor<
  u8[4][4]
  [
    [0, 1, 0, 0],
    [0, 0, 0, 0],
    [0, 0, 0, 0], 
    [0, 0, 0, 0] 
  ]
>
iex(6)> Nx.new_axis(y, 1) |> Nx.equal(x) |> Nx.any(axes: [0])
#Nx.Tensor<
  u8[4]
  [0, 1, 0, 0]
>
iex(7)> len_intersection = Nx.new_axis(y, 1) |> Nx.equal(x) |> Nx.any(axes: [0]) |> Nx.sum() 
#Nx.Tensor<
  u64
  1
>
iex(8)> len_union = Nx.size(x) |> Nx.add(Nx.size(y)) |> Nx.subtract(len_intersection)
#Nx.Tensor<
  s64
  7
>
iex(9)> Nx.divide(len_intersection, len_union)
#Nx.Tensor<
  f32
  0.1428571492433548
>
```

There might be some edge cases to explore, but this rationale of building the "equals table" at line `iex(5)` might help

**Comment by polvalente** [comment]
For instance, I don't know how this would behave for repeated values in one or both of the tensors. Might need to chose a different axis on `Nx.any` or change it to `Nx.sum`

**Comment by tiagodavi** [comment]
I noticed it's "hard" to implement simple things without the right data structure. Wouldn't be better if we could have sets natively as part of Nx? 

I was thinking to add something like this to Nx: https://scikit-learn.org/stable/modules/generated/sklearn.utils.multiclass.unique_labels.html 

To implement by using Tensors/Lists directly.   

I mean:

1 - Get unique elements from x1
2 - Get unique elements from x2
3 - Check the elements in common between x1 and x2 
4 - Join x1 + x2 as unique elements 
5 - Divide 

But I think having sets and this function above would be great.

**Comment by polvalente** [comment]
@tiagodavi I get what you mean. Currently a function like that isn't possible in Nx because the output shape can't be known at compile-time, and we don't support dynamic shapes yet.

I've experimented a bit and it seems that the following code works:

```elixir
iex> x
#Nx.Tensor<
  s64[9] 
  [1, 1, 3, 5, 7, 5, 5, 8, 9]
>
iex> y
#Nx.Tensor<
  s64[5] 
  [1, 2, 3, 4, 5]
>
iex> x_sorted = x |> Nx.sort()    
#Nx.Tensor<
  s64[9] 
  [1, 1, 3, 5, 5, 5, 7, 8, 9]
>

# This gets a tensor that represents a 1 whenever a given element is equal to the previous one,
# assuming no non-finite values are present.
iex> repeated_values_idx = Nx.conv(x_sorted |> Nx.new_axis(0) |> Nx.new_axis(0), Nx.tensor([[[-1, 1]]])) |> Nx.flatten() |> Nx.equal(0)
#Nx.Tensor<
  u8[8]
  [1, 0, 0, 1, 1, 0, 0, 0]
>
iex> len_intersection = x_sorted |> Nx.new_axis(1) |> Nx.equal(y) |> Nx.any(axes: [1]) |> Nx.logical_and(Nx.logical_not(Nx.concatenate([Nx.tensor([0]), repeated_values_idx]))) |> Nx.sum()
#Nx.Tensor<
  u64
  3
>
# This only takes into account repeated values in x. The same logic for repeated_values_idx would need to be calculated for y here.
# For the interesection calculation this isn't an issue, I think.
iex> len_union = Nx.size(x) |> Nx.add(Nx.size(y)) |> Nx.subtract(len_intersection) |> Nx.subtract(Nx.sum(repeated_values_idx))
#Nx.Tensor<
  s64
  8
>
iex> Nx.divide(len_intersection, len_union)
#Nx.Tensor<
  f32
  0.375
>
```

**Comment by jonatanklosko** [comment]
@polvalente instead of conv you can use element wise comparison with a shifted tensor (something like [this](https://github.com/jonatanklosko/meow/blob/2397468d09ac2f569d93f98f8781a6dd86a9e729/lib/meow_nx/utils.ex#L172)).

It would be best to benchmark the solutions, but the simples way I can think of right now is to:

* define `unique_size`  - basically sum over the mask I linked
* `x_size = unique_size(x)`
* `y_size = unique_size(y)`
* `union_size = unique_size(Nx.concatenate([x, y]))`
* `intersection_size = x_size + y_size - union_size`

I won't be able to test/measure this for w while, so just throwing the idea :)

**Comment by tiagodavi** [comment]
Thank you
